scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/nn/VflNNEstimator.scala

```diff
@@ -79,51 +88,84 @@ class VflNNEstimator(algorithm: String,
         val errors = getTensor("gradInput", gradInput)
         val loss = getTensor("loss", gradInput).value()
         model.backward(input, errors)
-
+        logger.debug(s"Model doing backward, version: $iteration")
         optimMethod.optimize(_ => (loss, grad), weight)
 
         iteration += 1
         count += miniBatch.size()
       }
       if (valDataSet != null) {
         model.evaluate()
-        val valIterator = valDataSet.data(false)
-        var evaluateResponse: EvaluateResponse = null;
-        while(valIterator.hasNext) {
-          val miniBatch = valIterator.next()
-          val input = miniBatch.getInput()
-          val target = miniBatch.getTarget()
-          val output = model.forward(input)
-          val metadata = TableMetaData.newBuilder
-            .setName(s"${model.getName()}_output").setVersion(iteration).build
-
-          // TODO: support table output and table target
-          val tableProto = outputTargetToTableProto(model.output, target, metadata)
-          evaluateResponse = flClient.nnStub.evaluate(tableProto, algorithm)
-        }
-        logger.info(evaluateResponse.getResponse)
-        val dataMap = evaluateResponse.getData.getTableMap.asScala
-        dataMap.foreach{v =>
-          if (evaluateResults.contains(v._1)) {
-            evaluateResults(v._1).append(v._2.getTensor(0))
-          } else {
-            evaluateResults(v._1) = ArrayBuffer(v._2.getTensor(0))
-          }
-        }
+        evaluate(valDataSet)
       }
 
     }
 
     model
   }
 
+  /**
+   * Evaluate VFL model
+   * For each batch, client estimator upload output tensor to server aggregator,
+   * the server keep updating the validation result until last batch of data,
+   * then return the evaluation result
+   * @param dataSet the data to evaluate
+   */
   override def evaluate(dataSet: LocalDataSet[MiniBatch[Float]]): Unit = {
-
-
+    val dataSize = dataSet.size()
+    val data = dataSet.data(false)
+    var count = 0
+    var iteration = 0
+    var evaluateResult = ""
+    while (count < dataSize) {
+      logger.debug(s"evaluating next batch, progress: $count/$dataSize")
+      model.evaluate()
+      val miniBatch = data.next()
+      val input = miniBatch.getInput()
+      val target = miniBatch.getTarget()
+      val output = model.forward(input)
+
+      val metadata = TableMetaData.newBuilder
+        .setName(s"${model.getName()}_output").setVersion(iteration).build
+      val tableProto = outputTargetToTableProto(model.output, target, metadata)
+      val hasReturn = if (!data.hasNext) true else false
+      evaluateResult = flClient.nnStub.evaluate(tableProto, algorithm, hasReturn).getMessage
+      iteration += 1
+      count += miniBatch.size()
+    }
+    println(s"Evaluation result: $evaluateResult")
```

Comment:
why print the last message only？

Reply:
This is the overall evaluation result string, although this may also not to be the best return. Any ideas?